streamflow/pystreamflow/model.py

```diff
@@ -2,41 +2,285 @@
 """
 
 import math
+import numpy as np
+import scipy.integrate
+import pints
+import pystreamflow
 
+# Try to import the optional scikit odes solver, and record whether or not it
+# was available
+try:
+    import scikits.odes
+    scikit_ode_avail = True
+except ImportError:
+    scikit_ode_avail = False
 
-def flux(S, a):
-    r"""Flux function for hydrological processes.
 
-    This is a general function for the relative flux as it depends on relative
-    storage and a single shape parameter.
+class RiverModel(pints.ForwardModel):
+    """Rainfall runoff model of river streamflow.
 
-    The equation is given by
+    The model has four latent state variables:
 
-    .. math::
-        f(S, a) = \frac{1 - e^{-aS}}{1 - e^{-a}}
+        * S_i = interception storage
+        * S_u = unsaturated storage
+        * S_s = slow reservoir
+        * S_f = fast reservoir
 
-    To prevent overflow, values of -a*S or -a which exceed 600 will be
-    truncated to 600. This limit should not be reached with typical values of a
-    and S, but it helps to protect the function during inference if strange
-    values are provided by some algorithm. Confer
-    https://github.com/Zaijab/DREAM/blob/master/examples/example_14/crr_model.c
+    and one observed variable:
 
-    Parameters
+        * z = water flowed out of the river
+
+    (In fact, it is dz/dt, or the streamflow, that is observed.)
+
+    The model is characterized by the following unknown parameters:
+
+        * I_max = maximum interception
+        * S_u,max = unsaturated storage capacity
+        * Q_s,max = maximum percolation
+        * alpha_e = evaporation flux shape
+        * alpha_f = runoff flux shape
+        * K_s = slow reservoir time constant
+        * K_f = fast reservoir time constant
+
+    as well as the following two parameters whose values are here assumed fixed
+    and known:
+
+        * alpha_s = 0 (percolation flux shape)
+        * alpha_i = 50 (interception flux shape)
+
+    Appearing multiple times in the model is the flux function f, given by:
+
+        * f(S, a) = (1 - exp(-a * S)) / (1 - exp(-a))
```

Comment:
I'd say these lines can be moved below the ODEs below since the increase the gap between there the state variables are introduced and where their dynamics are defined. Similarly, these parameters aren't actually used until someway below.